File: src/example_hdr.cpp
#include <nanogui/screen.h>
#include <nanogui/window.h>
#include <nanogui/layout.h>
#include <nanogui/label.h>
#include <nanogui/combobox.h>
#include <nanogui/imageview.h>
#include <nanogui/texture.h>
#include <nanogui/chroma.h>
#include <algorithm>
#include <cmath>
#include <vector>

using namespace nanogui;
using namespace nanogui::ituth273;

struct PrimaryEntry {
    ColorPrimaries primary;
    const char* name;
};

static const PrimaryEntry primary_table[] = {
    { ColorPrimaries::BT2020,     "BT2020" },
    { ColorPrimaries::BT470BG,    "BT470BG" },
    { ColorPrimaries::BT470M,     "BT470M" },
    { ColorPrimaries::BT709,      "BT709" },
    { ColorPrimaries::Film,       "Film" },
    { ColorPrimaries::SMPTE170M,  "SMPTE170M" },
    { ColorPrimaries::SMPTE240M,  "SMPTE240M" },
    { ColorPrimaries::SMPTE428,   "SMPTE428" },
    { ColorPrimaries::SMPTE431,   "SMPTE431" },
    { ColorPrimaries::SMPTE432,   "SMPTE432" }
};
static const int num_primaries = sizeof(primary_table) / sizeof(primary_table[0]);

class HDRGamutTest : public Screen {
public:
    HDRGamutTest() : Screen(Vector2i(800, 600), "NanoGUI test", /* resizable */ true, /* maximized */ false,
                            /* fullscreen */ false, /* depth_buffer */ true, /* stencil_buffer */ false,
                            /* float_buffer */ true, /* gl_major */ 3, /* gl_minor */ 2) {
        inc_ref();

        Window *window = new Window(this, "HDR & Color gamut test");
        window->set_layout(new GroupLayout());

        m_texture = new Texture(
            Texture::PixelFormat::RGBA,
            Texture::ComponentFormat::Float32,
            Vector2i(512, 400),
            Texture::InterpolationMode::Nearest,
            Texture::InterpolationMode::Nearest
        );

        new Label(window, "Primaries");

        std::vector<std::string> primary_names;
        for (int i = 0; i < num_primaries; ++i) {
            primary_names.push_back(primary_table[i].name);
        }

        ColorPrimaries screen_primary = from_screen(this);
        int primary_index = 0;
        for (int i = 0; i < num_primaries; ++i) {
            if (primary_table[i].primary == screen_primary) {
                primary_index = i;
                break;
            }
        }

        ComboBox *primaries_cbox = new ComboBox(window, primary_names);
        primaries_cbox->set_selected_index(primary_index);

        primaries_cbox->set_callback([this](int index) {
            auto chroma_array = chroma(primary_table[index].primary);
            m_rec709_matrix = chroma_to_rec709_matrix(chroma_array);
            update_texture();
        });

        // Initialize with current selection
        auto chroma_array = chroma(primary_table[primary_index].primary);
        m_rec709_matrix = chroma_to_rec709_matrix(chroma_array);

        new Label(window, "Linear ramps (0..4), bars mark integer values. Top: LDR sRGB, bottom: HDR with chosen primaries.");

        ImageView *img = new ImageView(window);
        img->set_image(m_texture);
        img->set_size(Vector2i(512, 400));
        img->set_scale(pixel_ratio());

        perform_layout();
        window->set_position(Vector2i(5, 5));

        update_texture();
    }

private:
    float to_srgb(float value) {
        float sign = value < 0 ? -1.0f : 1.0f;
        value = std::abs(value);
        return sign * (value < 0.0031308f ? value * 12.92f 
                                           : 1.055f * std::pow(value, 1.0f/2.4f) - 0.055f);
    }

    void update_texture() {
        const int width = 512, height = 400;
        std::vector<float> img(width * height * 4, 0.0f);

        // Draw vertical grid lines at quarter positions
        for (int y = 0; y < height; ++y) {
            for (int x : {0, width/4, width/2, 3*width/4, width-1}) {
                float* p = &img[(y * width + x) * 4];
                p[0] = p[1] = p[2] = p[3] = 1.0f;
            }
        }

        // Draw color bars: red, green, blue, white
        const struct { int y0, y1; Vector3f srgb, primary; } bars[] = {
            {10,  90,  {1, 0, 0}, m_rec709_matrix[0]},
            {110, 190, {0, 1, 0}, m_rec709_matrix[1]},
            {210, 290, {0, 0, 1}, m_rec709_matrix[2]},
            {310, 390, {1, 1, 1}, {1, 1, 1}}
        };

        for (auto& bar : bars) {
            int mid_y = (bar.y0 + bar.y1) / 2;
            for (int y = bar.y0; y < bar.y1; ++y) {
                bool is_srgb = (y < mid_y);
                for (int x = 0; x < width; ++x) {
                    float ramp = x * 4.0f / (width - 1);
                    float* p = &img[(y * width + x) * 4];
                    for (int c = 0; c < 3; ++c) {
                        float val = ramp * (is_srgb ? bar.srgb[c] : bar.primary[c]);
                        p[c] = to_srgb(is_srgb ? std::min(val, 1.0f) : val);
                    }
                    p[3] = 1.0f;
                }
            }
        }

        m_texture->upload((uint8_t*)img.data());
    }

private:
    ref<Texture> m_texture;
    Matrix3f m_rec709_matrix;
};

int main(int /* argc */, char ** /* argv */) {
    nanogui::init();

    {
        ref<HDRGamutTest> app = new HDRGamutTest();
        app->dec_ref();
        app->draw_all();
        app->set_visible(true);
        nanogui::run(RunMode::Lazy);
    }

    nanogui::shutdown();
    return 0;
}

File: src/python/example_hdr.py
import nanogui as ng
import numpy as np

class HDRGamutTest(ng.Screen):
    def __init__(self):
        super(HDRGamutTest, self).__init__(size=(800, 600), caption='NanoGUI test', float_buffer=True)
        window = ng.Window(self, 'HDR & Color gamut test')
        window.set_layout(ng.GroupLayout())

        self.texture = ng.Texture(
            pixel_format=ng.Texture.PixelFormat.RGBA,
            component_format=ng.Texture.ComponentFormat.Float32,
            min_interpolation_mode=ng.Texture.InterpolationMode.Nearest,
            mag_interpolation_mode=ng.Texture.InterpolationMode.Nearest,
            size=[512, 400],
        )

        ng.Label(window, "Primaries")
        primaries = sorted(ng.ituth273.ColorPrimaries, key=lambda x: x.__name__)
        primary_index = primaries.index(ng.ituth273.from_screen(self))
        primaries_cbox = ng.ComboBox(window, [p.__name__ for p in primaries])
        primaries_cbox.set_selected_index(primary_index)

        def primary_cb(index):
            p = ng.ituth273.chroma(primaries[index])
            rec709_matrix = ng.ituth273.chroma_to_rec709_matrix(p)
            self.rec709_matrix = np.from_dlpack(rec709_matrix)
            self.update_texture()

        primaries_cbox.set_callback(primary_cb)
        primary_cb(primary_index)

        ng.Label(window, "Linear ramps (0..4), bars mark integer values. Top: LDR sRGB, bottom: HDR with chosen primaries.")
        img = ng.ImageView(window)
        img.set_image(self.texture)
        img.set_size((512, 400))
        img.set_scale(self.pixel_ratio())

        self.perform_layout()
        window.set_position((5, 5))

    def create_color_bar(self, y_start, y_end, primary_color, srgb_color, max_value=4):
        img = np.zeros((y_end - y_start, 512, 4), dtype=np.float32)
        mid_y = (y_end - y_start) // 2

        ramp = np.linspace(0, max_value, 512)
        for i in range(3):
            img[:mid_y, :, i] = np.minimum(ramp * srgb_color[i], 1)
            img[mid_y:, :, i] = ramp * primary_color[i]
        img[..., 3] = 1

        return img

    def to_srgb(self, value):
        sign = np.sign(value)
        value = abs(value)
        return np.where(value < 0.0031308,
                        value * 12.92,
                        1.055 * value**(1.0/2.4) - 0.055) * sign

    def update_texture(self):
        """Update texture with color bars using rec709_matrix primaries and sRGB comparison"""
        img = np.zeros((400, 512, 4), dtype=np.float32)

        # Vertical grid lines
        img[:, 0*512//4, :] = img[:, 1*512//4, :] = 1
        img[:, 2*512//4, :] = img[:, 3*512//4, :] = 1
        img[:, 511, :] = 1

        bars = [
            (10, 90,   (1, 0, 0), self.rec709_matrix[:, 0]),
            (110, 190, (0, 1, 0), self.rec709_matrix[:, 1]),
            (210, 290, (0, 0, 1), self.rec709_matrix[:, 2]),
            (310, 390, (1, 1, 1), (1, 1, 1)),
        ]

        for y_start, y_end, srgb, primary in bars:
            img[y_start:y_end, :, :] = self.create_color_bar(y_start, y_end, primary, srgb)

        self.texture.upload(self.to_srgb(img))

def main():
    ng.init()
    app = HDRGamutTest()
    app.set_visible(True)
    ng.run(ng.RunMode.Lazy)
    del app
    ng.shutdown()

if __name__ == '__main__':
    main()
